concurrent_test.go

```diff
@@ -321,3 +384,100 @@ func testResultsDirectory(t *testing.T) string {
 
 	return path
 }
+
+/*
+*********************************************************
+Data structure and functions for analyzing history records
+*********************************************************
+*/
+type workerType string
+
+const (
+	reader workerType = "reader"
+	writer workerType = "writer"
+)
+
+type historyRecord struct {
+	WorkerType workerType `json:"workerType,omitempty"`
+	Key        string     `json:"key,omitempty"`
+	Value      []byte     `json:"value,omitempty"`
+	Txid       int        `json:"txid,omitempty"`
+}
+
+type historyRecords []historyRecord
+
+func (rs historyRecords) Len() int {
+	return len(rs)
+}
+
+func (rs historyRecords) Less(i, j int) bool {
+	// Sorted by key firstly: all records with the same key are grouped together.
+	keyCmp := strings.Compare(rs[i].Key, rs[j].Key)
+	if keyCmp < 0 {
+		return true
+	} else if keyCmp > 0 {
+		return false
+	}
+
+	// Sorted by txid
+	if rs[i].Txid < rs[j].Txid {
+		return true
+	} else if rs[i].Txid > rs[j].Txid {
+		return false
+	}
+
+	// Sorted by workerType: put writer before reader if they have the same txid.
+	if rs[i].WorkerType == writer {
+		return true
+	}
+
+	return false
+}
+
+func (rs historyRecords) Swap(i, j int) {
+	rs[i], rs[j] = rs[j], rs[i]
+}
+
+func collectHistoryRecords(recordingCh chan historyRecord) historyRecords {
+	var rs historyRecords
+	for record := range recordingCh {
+		rs = append(rs, record)
+	}
+	sort.Sort(rs)
+	return rs
+}
+
+func analyzeHistoryRecords(rs historyRecords) error {
```

Comment:
This is not a linearization validation as it doesn't account for operation execution time. It depends on Txid for order, making it very limited. For example it doesn't detect stale reads as View can just return old transation id, even though newer transaction was already committed. 

I'm not sure what concurrency issues this approach is mean to find or can find any. Please let me know if you need help with integrating linearizability checker https://github.com/anishathalye/porcupine like in robustness tests.

Reply:
Depending on Txid is enough, because in bboltDB txid is strictly incremental. Any data written by a transaction should be visible to following reading transaction, otherwise it must be an issue. My understanding is that porcupine might generate false alarm, but in this test it will never generate false alarm unless it's a test bug. Also we are testing only local boltDB library instead of a distributed system, so I don't see any reason to use porcupine.

Note that there is stale read, a read transaction should always see the same data during the transaction lifecycle, and it's exactly what repeatable reading means.

Reply:
In summary, there are two points:
1. A read transaction should always see the same data view during its lifecycle.
2. Any data written by a writing transaction should be visible to any following reading transactions (with txid >= previous writing txid)

Reply:
> You are mistaking serializability and linearizability.

Note that in bboltDB, only one writing transaction is allowed at a time. So in practice serializability is the same thing as linearizability.

Reply:
Goal of this tests is to drop all expectation and guesses, and let us validate them against the code behavior. Fact that it should not allow multiple writes, doesn't mean that we shouldn't test it.

Reply:
Note I am planning to add multiple writing transaction in followup PR.

Reply:
I can rename all `*linearizable*` to `*serializable*` when I start to support multiple writing transactions. thx for pointing this out.

Reply:
> I would recommend adding client side validation that transaction ID is always growing for each client, before we sort it.

Agreed. It's a good idea. Actually we also see such issue (txid isn't somehow strictly incremental). see https://github.com/etcd-io/bbolt/issues/402#issuecomment-1434220297

Let me do it in a separate PR. This PR is already too big.

Reply:
Ups, I mistook [Serializable](https://jepsen.io/consistency/models/serializable) vs [Sequential](https://jepsen.io/consistency/models/sequential). Still validating `sequential` here would be useful and ok to be done in followup.